### PR TITLE
refactor!: rename predicate to TupleSum/UnitSum

### LINKED ISSUE #448: Rename Predicate/SimplePredicate (to Choice + UnitChoice?)
I was thinking the same (BoundedUSize) but I take your argument. Really I think "predicate" is the wrong word to start. We're calling it that because we use it to control branches, but IMHO predicate is inherently binary - true or false - Google gives me this:

> 2.
> LOGIC
> something which is affirmed or denied concerning an argument of a proposition

So perhaps go with predicate for now but a better word might be "selector" or "choice"? or wilder ideas like "Nway"

_Originally posted by @acl-cqc in https://github.com/CQCL-DEV/hugr/pull/445#discussion_r1302914040_

### PR DESCRIPTION
**Pull request by ss2165**
Closes #448

BREAKING CHANGE: previous uses of type/value constructor methods for "predicates" will need updating

### CODE REVIEW

File: specification/hugr.md [hunk 1]
```diff
@@ -424,7 +424,7 @@ output of each of these is a sum type, whose arity is the number of outgoing
 control edges; the remaining outputs are those that are passed to all
 succeeding nodes.
 
-The three nodes labelled "Const" are simply generating a predicate with one empty
+The three nodes labelled "Const" are simply generating a choice with one empty
```
Comment (acl-cqc) [inline specification/hugr.md:427]: If a Choice (capitalized) is a type, is the corresponding value a choice (lowercase)? Or should "choice" here be "value of Choice type"? (the "one empty value to pass" might need changing too, not quite sure I understand that

File: specification/hugr.md [hunk 2]
```diff
@@ -335,16 +335,16 @@ express control flow, i.e. conditional or repeated evaluation.
...
 the constructor (tag) selects which child to execute; the unpacked
-contents of the Predicate with all remaining inputs to Conditional
+contents of the Choice with all remaining inputs to Conditional
 appended are sent to this child, and all outputs of the child are the
 outputs of the Conditional; that child is evaluated, but the others are
 not. That is, Conditional-nodes act as "if-then-else" followed by a
 control-flow merge.
 
-A **Predicate(T0, T1…TN)** type is an algebraic “sum of products” type,
+A **Choice(T0, T1…TN)** type is an algebraic “sum of products” type,
```
Comment (acl-cqc) [inline specification/hugr.md:347]: Drive-by, but I think we should be consistent between `T<i>` on this line and `#t<i>` on the next

File: src/builder/cfg.rs [hunk 3]
```diff
@@ -152,14 +152,14 @@ impl<B: AsMut<Hugr> + AsRef<Hugr>> CFGBuilder<B> {
...
-            predicate_variants,
+            choice_variants,
             other_outputs,
             inputs,
         )
     }
 
     /// Return a builder for a non-entry [`BasicBlock::DFB`] child graph with `inputs`
-    /// and `outputs` and a simple predicate type: a Sum of `n_cases` unit types.
+    /// and `outputs` and a unit choice type: a Sum of `n_cases` unit types.
```
Comment (acl-cqc) [inline src/builder/cfg.rs:162]:
```suggestion
    /// and `outputs` and a unit Choice type: a Sum of `n_cases` unit types.
```

File: src/builder/cfg.rs [hunk 4]
```diff
@@ -196,15 +196,15 @@ impl<B: AsMut<Hugr> + AsRef<Hugr>> CFGBuilder<B> {
...
+            choice_variants,
             other_outputs,
             extension_delta,
             true,
         )
     }
 
     /// Return a builder for the entry [`BasicBlock::DFB`] child graph with `inputs`
-    /// and `outputs` and a simple predicate type: a Sum of `n_cases` unit types.
+    /// and `outputs` and a unit choice type: a Sum of `n_cases` unit types.
```
Comment (acl-cqc) [inline src/builder/cfg.rs:207]:
```suggestion
    /// and `outputs` and a unit Choice type: a Sum of `n_cases` unit types.
```

File: src/builder/cfg.rs [hunk 5]
```diff
@@ -245,7 +245,7 @@ pub type BlockBuilder<B> = DFGWrapper<B, BasicBlockID>;
 
 impl<B: AsMut<Hugr> + AsRef<Hugr>> BlockBuilder<B> {
     /// Set the outputs of the block, with `branch_wire` being the value of the
-    /// predicate.  `outputs` are the remaining outputs.
+    /// Choice.  `outputs` are the remaining outputs.
```
Comment (acl-cqc) [inline src/builder/cfg.rs:248]: Why is it `branch_wire` not `choice_wire` (why wasn't it `predicate_wire` before)?

File: src/ops/constant.rs [hunk 6]
```diff
@@ -34,41 +34,41 @@ impl Const {
...
         }
     }
 
     /// Constant Sum over units, with only one variant.
-    pub fn simple_unary_predicate() -> Self {
+    pub fn unary_unit_choice() -> Self {
         Self {
-            value: Value::simple_unary_predicate(),
-            typ: Type::new_simple_predicate(1),
+            value: Value::unary_unit_choice(),
```
Comment (acl-cqc) [inline src/ops/constant.rs:59]: or just call `Self::unit_choice(0,1)`

File: src/ops/validate.rs [hunk 7]
```diff
@@ -146,18 +146,18 @@ impl ValidateOp for super::Conditional {
...
-        if self.predicate_inputs.len() != children.len() {
-            return Err(ChildrenValidationError::InvalidConditionalPredicate {
+        if self.choice_inputs.len() != children.len() {
+            return Err(ChildrenValidationError::InvalidConditionalChoice {
                 child: children[0].0, // Pass an arbitrary child
                 expected_count: children.len(),
-                actual_count: self.predicate_inputs.len(),
-                actual_predicate_rows: self.predicate_inputs.clone(),
+                actual_count: self.choice_inputs.len(),
+                actual_choice_rows: self.choice_inputs.clone(),
```
Comment (acl-cqc) [inline src/ops/validate.rs:156]: just a side note, elsewhere this would be "choice_variants". I wonder whether "choice_rows" would be better everywhere

File: src/types.rs [hunk 8]
```diff
@@ -241,17 +241,17 @@ impl Type {
...
     where
         V: Into<TypeRow>,
     {
-        Self::new_sum(predicate_variants_row(variant_rows))
+        Self::new_sum(choice_variant_row(variant_rows))
     }
 
-    /// New simple predicate with empty Tuple variants
-    pub const fn new_simple_predicate(size: u8) -> Self {
+    /// New unit choice with empty Tuple variants
```
Comment (acl-cqc) [inline src/types.rs:253]:
```suggestion
    /// New unit Choice with empty Tuple variants
```

File: src/types.rs [hunk 9]
```diff
@@ -142,7 +142,7 @@ impl SumType {
 impl From<SumType> for Type {
     fn from(sum: SumType) -> Type {
         match sum {
-            SumType::Simple { size } => Type::new_simple_predicate(size),
+            SumType::Simple { size } => Type::new_unit_choice(size),
```
Comment (acl-cqc) [inline src/types.rs:145]: Reckon you should rename SumType::Simple. Maybe SumType::OfUnits ? (SumType::Unit perhaps but seems more likely to lead to confusion)

File: src/types/serialize.rs [hunk 10]
```diff
@@ -80,8 +80,8 @@ mod test {
         let t = Type::new_sum(vec![USIZE_T, FLOAT64_TYPE]);
         assert_eq!(ser_roundtrip(&t), t);
 
-        // A simple predicate
-        let t = Type::new_simple_predicate(4);
+        // A unit choice
```
Comment (acl-cqc) [inline src/types/serialize.rs:83]:
```suggestion
        // A unit Choice
```
But I don't think this comment adds much tbh....(!)

File: src/values.rs [hunk 11]
```diff
@@ -96,14 +96,14 @@ impl Value {
         Self::Tuple { vs: vec![] }
     }
 
-    /// Constant Sum over units, used as predicates.
-    pub fn simple_predicate(tag: usize) -> Self {
+    /// Constant Sum over units, used as branching values.
```
Comment (acl-cqc) [inline src/values.rs:99]:
```suggestion
    /// Constant Sum of a unit value, used to control branches.
```
(Other variants could be non-unit, since we are checkable not synthesizable here)

File: src/values.rs [hunk 12]
```diff
@@ -96,14 +96,14 @@ impl Value {
...
+    pub fn unit_choice(tag: usize) -> Self {
         Self::sum(tag, Self::unit())
     }
 
     /// Constant Sum over Tuples with just one variant of unit type
```
Comment (acl-cqc) [inline src/values.rs:104]:
```suggestion
    /// Constant Sum with just one variant of unit type
```

File: src/values.rs [hunk 13]
```diff
@@ -113,7 +113,7 @@ impl Value {
         }
     }
 
-    /// Sum value (could be of any compatible type, e.g. a predicate)
+    /// Sum value (could be of any compatible type, e.g. a Choice)
```
Comment (acl-cqc) [inline src/values.rs:116]:
```suggestion
    /// Sum value (could be of any compatible type - e.g., if `value` was a Tuple, a Choice type)
```

**Review by acl-cqc — COMMENTED** [review]
Yes-ish. I think I'd like @cqc-alec's looking over here, not for nitpicking details, but whether this name-change "works" overall.
* "Unit Choice" is clearly an improvement over "Simple Predicate", but then "Unit Predicate" would have been too
* My general feeling is I'm not as persuaded as I thought I'd be. I guess I just feel that one is far more likely to use the word choice by accident, when one didn't mean a Choice, and in spoken English capitalization is hard to determine.
* Perhaps this is OK if we are 100% consistent on always capitalizing here, and it may also help to tighten up the distinction between Choice types and Choice values (if that's what they are called) - by which perhaps I mean never to use "Choice" on its own without being immediately followed by either "type" or "value"...
* I do like "choice_variants" or "choice_rows", too. Perhaps a general theme here that maybe Choice works better if we always use it as an adjective not as a noun?

I'm not sure whether I see a good alternative that is less likely to be used by accident. "Selector" perhaps, but is that better than Predicate? "Chooser"? Partway through I was actually wondering if we could just say "sum-of-tuples" explicitly, but then I gave up when I looked at the builder code....

**Comment by cqc-alec** [comment]
> Yes-ish. I think I'd like @cqc-alec's looking over here, not for nitpicking details, but whether this name-change "works" overall.
> 
>     * "Unit Choice" is clearly an improvement over "Simple Predicate", but then "Unit Predicate" would have been too

Or "Simple Choice"?

>     * My general feeling is I'm not as persuaded as I thought I'd be. I guess I just feel that one is far more likely to use the word choice by accident, when one didn't mean a Choice, and in spoken English capitalization is hard to determine.
> 
>     * Perhaps this is OK if we are 100% consistent on always capitalizing here, and it may also help to tighten up the distinction between Choice types and Choice values (if that's what they are called) - by which perhaps I mean never to use "Choice" on its own without being immediately followed by either "type" or "value"...
> 
>     * I do like "choice_variants" or "choice_rows", too. Perhaps a general theme here that maybe Choice works better if we always use it as an adjective not as a noun?
> 
> 
> I'm not sure whether I see a good alternative that is less likely to be used by accident. "Selector" perhaps, but is that better than Predicate? "Chooser"? Partway through I was actually wondering if we could just say "sum-of-tuples" explicitly, but then I gave up when I looked at the builder code....

"Choice" seems better than "predicate" but perhaps it isn't ideal. We could use "variant" (variant type or variant value), as in C++17. That can be heard as a noun or adjective, but I'm not sure if that ambiguity matters. I'm less keen on "chooser" or "selector" because they imply an _action_ that is not really part of the type.

**Comment by ss2165** [comment]
what about borrowing SOP (sum of products) from boolean algebra?

**Comment by cqc-alec** [comment]
> what about borrowing SOP (sum of products) from boolean algebra?

Has the advantage of being precise, but I feel like a more understandable name (for exactly that concept) would be preferable. If "Choice" and "Variant" are a bit too vague, could we use "DataChoice" or "DataVariant"? Not really convinced, just trying out ideas...

**Comment by ss2165** [comment]
Variant is too close to Sum for me (indeed we use it in Tierkreis instead of Sum). How about just simple `SumTuple` and `SumUnit` then?
`new_sum_tuple`, `new_sum_unit` etc.
We could keep using Choice just as a shorthand in the spec, and use precise naming in the code.

**Comment by cqc-alec** [comment]
> Variant is too close to Sum for me (indeed we use it in Tierkreis instead of Sum). How about just simple `SumTuple` and `SumUnit` then? `new_sum_tuple`, `new_sum_unit` etc. We could keep using Choice just as a shorthand in the spec, and use precise naming in the code.

I'd personally find it confusing to have different terminology used in the spec and implementation. I'd be OK with something like `SumTuple` and `SumUnit`, though I would turn these around to `TupleSum` and `UnitSum`!

**Comment by acl-cqc** [comment]
If `TupleSum` seems ok (I guess it's no longer than predicate!) then that could work. I'm wondering if "choice value" and "choice type" might be alright tho - specifically, if we always use choice as an adjective, then that's different from its normal/accidental use in English, i.e. as a noun. Perhaps `Type::new_choice` is obviously-enough a new choice *type* to avoid having to write `Type::new_choice_type`...

**Comment by ss2165** [comment]
Yeah I see what you mean Alan, `Choice` reads nicer in the spec, but I'm happy with `TupleSum` and `UnitSum` everywhere

**Comment by cqc-alec** [comment]
> Yeah I see what you mean Alan, `Choice` reads nicer in the spec, but I'm happy with `TupleSum` and `UnitSum` everywhere

OK for me

**Comment by cqc-alec** [comment]
Sorry closed by mistake!

**Review by cqc-alec — COMMENTED** [review]
Do we go with `TupleSum` and `UnitSum` then?

File: src/ops/validate.rs (hunk 7 above)
Comment (ss2165) [reply src/ops/validate.rs]: yes I've tried to go with "tuple_sum_rows" everywhere

File: src/extension/infer.rs [hunk 14]
```diff
@@ -1209,11 +1210,11 @@ mod test {
         entry_extensions: ExtensionSet,
         exit_types: impl Into<TypeRow>,
     ) -> Result<([Node; 3], Node), Box<dyn Error>> {
-        let entry_predicate_type = Type::new_predicate(entry_predicates.clone());
+        let entry_predicate_type = Type::new_tuple_sum(entry_predicates.clone());
```
Comment (cqc-alec) [inline src/extension/infer.rs:1213]: Should we rename `entry_predicates`?

**Review by cqc-alec — APPROVED** [review]
